libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c

```diff
@@ -0,0 +1,3476 @@
+
+				if( xBool != CK_TRUE )
+				{
+					PKCS11_PRINT( ( "ERROR: Only keys with signing priveledges are supported. \r\n" ) );
+					xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+				}
+
+				break;
+
+			case ( CKA_LABEL ):
+
+				if( xAttribute.ulValueLen <= pkcs11configMAX_LABEL_LENGTH )
+				{
+					*ppxLabel = &pxTemplate[ ulIndex ];
+				}
+				else
+				{
+					xResult = CKR_DATA_LEN_RANGE;
+				}
+
+				break;
+
+			case ( CKA_EC_PARAMS ):
+
+				if( memcmp( ( CK_BYTE[] ) pkcs11DER_ENCODED_OID_P256, xAttribute.pValue, xAttribute.ulValueLen ) )
+				{
+					PKCS11_PRINT( ( "ERROR: Only elliptic curve P-256 is supported.\r\n" ) );
+					xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+				}
+
+				break;
+
+			case ( CKA_VALUE ):
+				lMbedReturn = mbedtls_mpi_read_binary( &pxKeyPair->d,
+													   xAttribute.pValue,
+													   xAttribute.ulValueLen );
+
+				if( lMbedReturn != 0 )
+				{
+					xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+				}
+
+				break;
+
+			default:
+				xResult = CKR_TEMPLATE_INCONSISTENT;
+				break;
+		}
+	}
+
+	return xResult;
+}
+
+/* Helper function for parsing RSA Private Key attribute templates
+ * for C_CreateObject. */
+CK_RV prvCreateRsaPrivateKey( mbedtls_pk_context * pxMbedContext,
+							  CK_ATTRIBUTE_PTR * ppxLabel,
+							  CK_ATTRIBUTE_PTR * ppxClass,
+							  CK_ATTRIBUTE_PTR pxTemplate,
+							  CK_ULONG ulCount,
+							  CK_OBJECT_HANDLE_PTR pxObject )
+{
+	CK_RV xResult = CKR_OK;
+	mbedtls_rsa_context * pxRsaContext;
+	int lMbedReturn = 0;
+	CK_BBOOL xBool;
+	uint32_t ulIndex;
+	CK_ATTRIBUTE xAttribute;
+
+	*ppxLabel = NULL;
+	*ppxClass = NULL;
+	pxRsaContext = pxMbedContext->pk_ctx;
+	mbedtls_rsa_init( pxRsaContext, MBEDTLS_RSA_PKCS_V15, 0 /*ignored.*/ );
+
+	/* Parse template and collect the relevant parts. */
+	for( ulIndex = 0; ulIndex < ulCount; ulIndex++ )
+	{
+		xAttribute = pxTemplate[ ulIndex ];
+
+		switch( xAttribute.type )
+		{
+			case ( CKA_CLASS ):
+				*ppxClass = &pxTemplate[ ulIndex ];
+				break;
+
+			case ( CKA_KEY_TYPE ):
+
+				/* Do nothing.
+				 * Key type & object type were checked previously.
+				 */
+				break;
+
+			case ( CKA_TOKEN ):
+				memcpy( &xBool, xAttribute.pValue, sizeof( CK_BBOOL ) );
+
+				if( xBool != CK_TRUE )
+				{
+					PKCS11_PRINT( ( "ERROR: Only token key creation is supported. \r\n" ) );
+					xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+				}
+
+				break;
+
+			case ( CKA_LABEL ):
+
+				if( xAttribute.ulValueLen <= pkcs11configMAX_LABEL_LENGTH )
+				{
+					*ppxLabel = &pxTemplate[ ulIndex ];
+				}
+				else
+				{
+					xResult = CKR_DATA_LEN_RANGE;
+				}
+
+				break;
+
+			case ( CKA_SIGN ):
+				memcpy( &xBool, xAttribute.pValue, xAttribute.ulValueLen );
+
+				if( xBool != CK_TRUE )
+				{
+					PKCS11_PRINT( ( "Only RSA private keys with signing permissions supported. \r\n" ) );
+					xResult = CKR_TEMPLATE_INCONSISTENT;
+				}
+
+				break;
+
+			case ( CKA_MODULUS ):
+				lMbedReturn |= mbedtls_rsa_import_raw( pxRsaContext,
```

Comment:
Why OR this in?

Reply:
The intention is to gather the errors generated in checking the necessary components of the key. At the end of the function, the CKR_ATTRIBUTE_VALUE_INVALID error is returned if any error occurred during the check.

Reply:
Make sense. I will update it in the next patch.